File: engine/timeline.py

#include <vector>
#include "clip.h"


class TimelineClip
{
    public:
        TimelineClip();
        Frame* time_to_frame(Time t);
        void set_clip(Clip* c);
        void set_zone(Zone* z);
        void set_start(Time t);
        void set_stop(Time t);

    private:
        Clip* clip;
        Zone* zone;
        Time start;
        Time stop;        
};

class Timeline
{
    public:
        Timeline();
        Frame* time_to_frame(Time t);
        void add_clip(TimelineClip* c);
        void remove_clip(TimelineClip* c);
    private:
        std::vector<TimelineClip*> clips;

};

File: lzr_engine/src/timeline.cpp

#include <timeline.h>


TimelineClip::TimelineClip()
{

}

TimelineClip::~TimelineClip()
{

}

lzr_frame TimelineClip::render(lzr_time t)
{
    //teh frame buffer used during rendering
    lzr_frame frame;
    frame.n_points = 0;

    //if the requested time falls within this clip
    if((start >= t) && (t <= stop))
    {
        //normalize the timecode, according to the clip's start and stop times
        t = (t - start) / (stop - start);
        //render the clip
        clip->render(&frame, t);
    }

    return frame;
}

void TimelineClip::set_clip(Clip* c)
{
    clip = c;
}

void TimelineClip::set_start(Time t)
{
    start = t;
}

void TimelineClip::set_stop(Time t)
{
    stop = t;
}
